tests/cli/test_deploy.py

```diff
@@ -2397,6 +2397,84 @@ async def test_deploy_does_not_prompt_build_docker_image_when_empty_build_action
             "An important name/test-name"
         )
 
+    @pytest.mark.usefixtures("project_dir", "interactive_console")
+    async def test_deploy_templates_env_vars(self, prefect_client, monkeypatch):
+        # work pool must exist
+        await prefect_client.create_work_pool(
+            WorkPoolCreate(name="test-pool", type="test")
+        )
+        # set up environment variables
+        monkeypatch.setenv("WORK_POOL", "test-pool")
+        monkeypatch.setenv("MY_VAR", "my-value")
+        monkeypatch.setenv("REGISTRY", "my-registry")
+        monkeypatch.setenv("IMAGE_NAME", "my-image")
+        monkeypatch.setenv("TAG", "my-tag")
+
+        # set up prefect.yaml that has env var placeholders in a variety of places
+        prefect_file = Path("prefect.yaml")
+        with prefect_file.open(mode="r") as f:
+            prefect_config = yaml.safe_load(f)
+        prefect_config["definitions"] = [
+            {
+                "work_pools": {
+                    "work_pool": "&work_pool",
+                    "name": "{{ $WORK_POOL_NAME}}",
+                    "job_variables": {
+                        "image": "{{ $REGISTRY }}/{{ $IMAGE_NAME }}:{{ $TAG }}"
+                    },
+                }
+            }
+        ]
+        prefect_config["deployments"] = [
+            {
+                "name": "test-deployment",
+                "entrypoint": "flows/hello.py:my_flow",
+                "work_pool": "*work_pool",
+            }
+        ]
+        prefect_config["build"] = [
+            {"prefect.testing.utilities.a_test_step": {"input": "{{ $MY_VAR }}"}}
+        ]
+
+        # save config to prefect.yaml
+        with prefect_file.open(mode="w") as f:
+            yaml.safe_dump(prefect_config, f)
+
+        result = await run_sync_in_worker_thread(
+            invoke_and_assert,
+            command="deploy ./flows/hello.py:my_flow -n test-name -p test-pool",
+            expected_code=0,
+            expected_output_contains="An important name/test-name",
+            user_input=(
+                "n"  # Decline schedule
+                + readchar.key.ENTER
+                + "n"  # Decline docker build
+                + readchar.key.ENTER
+                + "y"  # Accept save configuration
+                + readchar.key.ENTER
+            ),
+        )
+        assert result.exit_code == 0
+        assert "An important name/test" in result.output
+
+        deployment = await prefect_client.read_deployment_by_name(
+            "An important name/test-name"
+        )
+        assert deployment.work_pool_name == "test-pool"
```

Comment:
Just to be sure, can we also check the env vars in `infra_overrides` are templated?
```suggestion
        assert deployment.work_pool_name == "test-pool"
        assert deployment.infra_overrides == { "image": "my-registry/my-image:my-tag" }
```

Reply:
@desertaxle I've updated the test in [16b6109](https://github.com/PrefectHQ/prefect/pull/11463/commits/16b6109bc687bfcf39cf6ad87609eef4eebf4c16)

Reply:
And I can confirm this test would fail without the code change in this PR